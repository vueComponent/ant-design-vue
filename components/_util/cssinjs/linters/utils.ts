import devWarning from '../../../vc-util/warning';
import type { LinterInfo } from './interface';

export function lintWarning(message: string, info: LinterInfo) {
  const { path, parentSelectors } = info;

  devWarning(
    false,
    `[Ant Design Vue CSS-in-JS] ${path ? `Error in '${path}': ` : ''}${message}${
      parentSelectors.length ? ` Selector info: ${parentSelectors.join(' -> ')}` : ''
    }`,
  );
}
