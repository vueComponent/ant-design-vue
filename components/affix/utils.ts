import addEventListener from '../vc-util/Dom/addEventListener';
import supportsPassive from '../_util/supportsPassive';

export type BindElement = HTMLElement | Window | null | undefined;

export function getTargetRect(target: BindElement): DOMRect {
  return target !== window
    ? (target as HTMLElement).getBoundingClientRect()
    : ({ top: 0, bottom: window.innerHeight } as DOMRect);
}

export function getFixedTop(placeholderRect: DOMRect, targetRect: DOMRect, offsetTop: number) {
  if (offsetTop !== undefined && targetRect.top > placeholderRect.top - offsetTop) {
    return `${offsetTop + targetRect.top}px`;
  }
  return undefined;
}

export function getFixedBottom(
  placeholderRect: DOMRect,
  targetRect: DOMRect,
  offsetBottom: number,
) {
  if (offsetBottom !== undefined && targetRect.bottom < placeholderRect.bottom + offsetBottom) {
    const targetBottomOffset = window.innerHeight - targetRect.bottom;
    return `${offsetBottom + targetBottomOffset}px`;
  }
  return undefined;
}

// ======================== Observer ========================
const TRIGGER_EVENTS = [
  'resize',
  'scroll',
  'touchstart',
  'touchmove',
  'touchend',
  'pageshow',
  'load',
];

interface ObserverEntity {
  target: HTMLElement | Window;
  affixList: any[];
  eventHandlers: { [eventName: string]: any };
}

let observerEntities: ObserverEntity[] = [];

export function getObserverEntities() {
  // Only used in test env. Can be removed if refactor.
  return observerEntities;
}

export function addObserveTarget<T>(target: HTMLElement | Window | null, affix: T): void {
  if (!target) return;

  let entity = observerEntities.find(item => item.target === target);

  if (entity) {
    entity.affixList.push(affix);
  } else {
    entity = {
      target,
      affixList: [affix],
      eventHandlers: {},
    };
    observerEntities.push(entity);

    // Add listener
    TRIGGER_EVENTS.forEach(eventName => {
      entity!.eventHandlers[eventName] = addEventListener(target, eventName, () => {
        entity!.affixList.forEach(
          targetAffix => {
            const { lazyUpdatePosition } = (targetAffix as any).exposed;
            lazyUpdatePosition();
          },
          (eventName === 'touchstart' || eventName === 'touchmove') && supportsPassive
            ? ({ passive: true } as EventListenerOptions)
            : false,
        );
      });
    });
  }
}

export function removeObserveTarget<T>(affix: T): void {
  const observerEntity = observerEntities.find(oriObserverEntity => {
    const hasAffix = oriObserverEntity.affixList.some(item => item === affix);
    if (hasAffix) {
      oriObserverEntity.affixList = oriObserverEntity.affixList.filter(item => item !== affix);
    }
    return hasAffix;
  });

  if (observerEntity && observerEntity.affixList.length === 0) {
    observerEntities = observerEntities.filter(item => item !== observerEntity);

    // Remove listener
    TRIGGER_EVENTS.forEach(eventName => {
      const handler = observerEntity.eventHandlers[eventName];
      if (handler && handler.remove) {
        handler.remove();
      }
    });
  }
}
