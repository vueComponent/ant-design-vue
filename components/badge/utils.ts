import { PresetColorTypes, PresetStatusColorTypes } from '../_util/colors';

export function isPresetColor(color?: string): boolean {
  return (
    (PresetColorTypes as any[]).indexOf(color) !== -1 ||
    (PresetStatusColorTypes as any[]).indexOf(color) !== -1
  );
}
