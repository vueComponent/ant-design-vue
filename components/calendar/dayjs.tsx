import generateConfig from '../vc-picker/generate/dayjs';
import type { CalendarProps } from './generateCalendar';
import generateCalendar from './generateCalendar';

const Calendar = generateCalendar(generateConfig);

export type { CalendarProps };
export default Calendar;
