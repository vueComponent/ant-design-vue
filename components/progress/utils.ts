import { presetPrimaryColors } from '@ant-design/colors';
import devWarning from '../vc-util/devWarning';
import type { CircleProps } from './Circle';
import type { ProgressProps } from './props';

export function validProgress(progress: number | undefined) {
  if (!progress || progress < 0) {
    return 0;
  }
  if (progress > 100) {
    return 100;
  }
  return progress;
}

export function getSuccessPercent({ success, successPercent }: ProgressProps) {
  let percent = successPercent;
  /** @deprecated Use `percent` instead */
  if (success && 'progress' in success) {
    devWarning(
      false,
      'Progress',
      '`success.progress` is deprecated. Please use `success.percent` instead.',
    );
    percent = success.progress;
  }
  if (success && 'percent' in success) {
    percent = success.percent;
  }
  return percent;
}

export function getPercentage({ percent, success, successPercent }: ProgressProps) {
  const realSuccessPercent = validProgress(getSuccessPercent({ success, successPercent }));
  return [realSuccessPercent, validProgress(validProgress(percent) - realSuccessPercent)];
}

export function getStrokeColor({
  success = {},
  strokeColor,
}: Partial<CircleProps>): (string | Record<string, string>)[] {
  const { strokeColor: successColor } = success;
  return [successColor || presetPrimaryColors.green, strokeColor || null!];
}
