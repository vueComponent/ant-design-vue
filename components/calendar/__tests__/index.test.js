import dayjs from 'dayjs';
import { mount } from '@vue/test-utils';
import { asyncExpect, sleep } from '../../../tests/utils';
import MockDate from 'mockdate';
import Calendar from '..';
import Header from '../Header';
import mountTest from '../../../tests/shared/mountTest';
import generateConfig from '../../vc-picker/generate/dayjs';

describe('Calendar', () => {
  mountTest(Calendar);
  beforeEach(() => {
    document.body.innerHTML = '';
  });

  function openSelect(wrapper, className) {
    wrapper.find(className).find('.ant-select-selector').trigger('mousedown');
  }

  function findSelectItem(wrapper) {
    return wrapper.findAll('.ant-select-item-option');
  }

  function clickSelectItem(wrapper, index = 0) {
    findSelectItem(wrapper)[index].trigger('click');
  }

  it('Calendar should be selectable', async () => {
    const onSelect = jest.fn();
    const wrapper = mount(
      {
        render() {
          return <Calendar onSelect={onSelect} />;
        },
      },
      { sync: false },
    );
    await asyncExpect(() => {
      wrapper.findAll('.ant-picker-cell')[0].trigger('click');
    }, 0);
    await asyncExpect(() => {
      expect(onSelect).toHaveBeenCalled();
      const value = onSelect.mock.calls[0][0];
      expect(dayjs.isDayjs(value)).toBe(true);
    });
  });

  it('only Valid range should be selectable', async () => {
    const onSelect = jest.fn();
    const validRange = [dayjs('2018-02-02'), dayjs('2018-02-18')];
    const wrapper = mount(
      {
        render() {
          return (
            <Calendar
              onSelect={onSelect}
              validRange={validRange}
              defaultValue={dayjs('2018-02-02')}
            />
          );
        },
      },
      { sync: false },
    );
    await asyncExpect(() => {
      wrapper.findAll('[title="2018-02-01"]')[0].trigger('click');
      wrapper.findAll('[title="2018-02-02"]')[0].trigger('click');
      expect(onSelect.mock.calls.length).toBe(1);
    });
  });

  it('dates other than in valid range should be disabled', async () => {
    const onSelect = jest.fn();
    const validRange = [dayjs('2018-02-02'), dayjs('2018-02-18')];
    const wrapper = mount(
      {
        render() {
          return (
            <Calendar
              onSelect={onSelect}
              validRange={validRange}
              defaultValue={dayjs('2018-02-02')}
            />
          );
        },
      },
      { sync: false },
    );
    await asyncExpect(() => {
      wrapper.findAll('[title="2018-02-20"]')[0].trigger('click');
      expect(wrapper.find('[title="2018-02-20"]').classes()).toContain('ant-picker-cell-disabled');
      expect(onSelect.mock.calls.length).toBe(0);
    });
  });

  it('months other than in valid range should be disabled', async () => {
    const onSelect = jest.fn();
    const validRange = [dayjs('2018-02-02'), dayjs('2018-05-18')];
    const wrapper = mount(
      {
        render() {
          return (
            <Calendar
              onSelect={onSelect}
              validRange={validRange}
              defaultValue={dayjs('2018-02-02')}
              mode="year"
            />
          );
        },
      },
      { sync: false },
    );
    await asyncExpect(() => {
      expect(wrapper.findAll('[title="2018-01"]')[0].classes()).toContain(
        'ant-picker-cell-disabled',
      );
      expect(wrapper.findAll('[title="2018-02"]')[0].classes()).not.toContain(
        'ant-picker-cell-disabled',
      );
      expect(wrapper.findAll('[title="2018-06"]')[0].classes()).toContain(
        'ant-picker-cell-disabled',
      );
      wrapper.findAll('[title="2018-01"]')[0].trigger('click');
      wrapper.findAll('[title="2018-03"]')[0].trigger('click');
      expect(onSelect.mock.calls.length).toBe(1);
    });
  });

  it('months other than in valid range should not be shown in header', async () => {
    document.body.innerHTML = '';
    const validRange = [dayjs('2017-02-02'), dayjs('2018-05-18')];
    // eslint-disable-next-line no-unused-vars
    const wrapper = mount(
      {
        render() {
          return <Calendar validRange={validRange} />;
        },
      },
      { sync: false, attachTo: 'body' },
    );
    await sleep();
    openSelect(wrapper, '.ant-picker-calendar-year-select');
    await sleep(100);
    clickSelectItem(wrapper);
    await sleep();
    openSelect(wrapper, '.ant-picker-calendar-month-select');
    await sleep(100);
    // 2 years and 11 months
    expect(wrapper.findAll('.ant-select-item-option').length).toBe(13);
  });

  it('getDateRange should returns a disabledDate function', async () => {
    const validRange = [dayjs('2018-02-02'), dayjs('2018-05-18')];
    const wrapper = mount(Calendar, {
      props: {
        validRange,
        defaultValue: dayjs('2018-02-02'),
      },
      sync: false,
    });
    await asyncExpect(() => {
      const { disabledDate } = wrapper.getComponent({ name: 'PickerPanel' }).props();
      expect(disabledDate(dayjs('2018-06-02'))).toBe(true);
      expect(disabledDate(dayjs('2018-04-02'))).toBe(false);
    });
  });

  it('Calendar should change mode by prop', async () => {
    const monthMode = 'month';
    const yearMode = 'year';
    const wrapper = mount(Calendar, { sync: false });
    await sleep(50);
    expect(wrapper.getComponent({ name: 'CalendarHeader' }).props().mode).toEqual(monthMode);
    wrapper.setProps({ mode: 'year' });
    await sleep(50);
    expect(wrapper.getComponent({ name: 'CalendarHeader' }).props().mode).toEqual(yearMode);
  });

  it('Calendar should switch mode', async () => {
    const monthMode = 'month';
    const yearMode = 'year';
    const onPanelChangeStub = jest.fn();
    const wrapper = mount(Calendar, {
      props: {
        mode: yearMode,
        onPanelChange: onPanelChangeStub,
      },
      sync: false,
    });
    await asyncExpect(() => {
      expect(wrapper.getComponent({ name: 'CalendarHeader' }).props().mode).toEqual(yearMode);
      wrapper.setProps({ mode: monthMode });
    });
    await asyncExpect(() => {
      expect(wrapper.getComponent({ name: 'CalendarHeader' }).props().mode).toEqual(monthMode);
      expect(onPanelChangeStub).toHaveBeenCalledTimes(0);
    });
  });

  it('Calendar should support locale', async () => {
    MockDate.set(dayjs('2018-10-19'));
    // eslint-disable-next-line
    const zhCN = require('../locale/zh_CN').default;
    const wrapper = mount(Calendar, {
      props: {
        locale: zhCN,
      },
      sync: false,
    });
    await asyncExpect(() => {
      expect(wrapper.html()).toMatchSnapshot();
      MockDate.reset();
    });
  });

  it('should trigger onPanelChange when click last month of date', () => {
    const onPanelChange = jest.fn();
    const date = new dayjs('1990-09-03');
    const wrapper = mount(Calendar, {
      props: {
        value: date,
        onPanelChange,
      },
      sync: false,
    });
    wrapper.findAll('.ant-picker-cell')[0].trigger('click');

    expect(onPanelChange).toHaveBeenCalled();
    expect(onPanelChange.mock.calls[0][0].month()).toEqual(date.month() - 1);
  });

  it('switch should work correctly without prop mode', async () => {
    const onPanelChange = jest.fn();
    const date = new dayjs(new Date(Date.UTC(2017, 7, 9, 8)));
    const wrapper = mount(Calendar, {
      props: {
        value: date,
        onPanelChange,
      },
      sync: false,
      attachTo: 'body',
    });
    await sleep(300);
    expect(wrapper.getComponent({ name: 'CalendarHeader' }).props().mode).toBe('month');
    expect(wrapper.findAll('.ant-picker-date-panel').length).toBe(1);
    expect(wrapper.findAll('.ant-picker-month-panel').length).toBe(0);
    await wrapper.findAll('.ant-radio-button-input[value="year"]')[0].trigger('change');
    await sleep(300);
    expect(wrapper.findAll('.ant-picker-date-panel').length).toBe(0);
    expect(wrapper.findAll('.ant-picker-month-panel').length).toBe(1);
    expect(onPanelChange).toHaveBeenCalled();
    expect(onPanelChange.mock.calls[0][1]).toEqual('year');
  });

  // const createWrapper = async (start, end, value, onValueChange) => {
  //   document.body.innerHTML = '';
  //   const wrapper = mount(
  //     {
  //       render() {
  //         return (
  //           <Header
  //             prefixCls="ant-picker-calendar"
  //             onChange={onValueChange}
  //             generateConfig={generateConfig}
  //             value={value}
  //             validRange={[start, end]}
  //             locale={{ year: 'å¹´' }}
  //           />
  //         );
  //       },
  //     },
  //     {
  //       sync: false,
  //       attachTo: 'body',
  //     },
  //   );
  //   await sleep(50);
  //   openSelect(wrapper, '.ant-picker-calendar-year-select');
  //   await sleep(50);
  //   clickSelectItem(wrapper);
  //   await sleep(50);
  // };

  // it('if value.month > end.month, set value.month to end.month', async () => {
  //   const value = new dayjs('1990-01-03');
  //   const start = new dayjs('2019-04-01');
  //   const end = new dayjs('2019-11-01');
  //   const onValueChange = jest.fn();
  //   await createWrapper(start, end, value, onValueChange);
  //   expect(onValueChange).toHaveBeenCalledWith(value.year('2019').month('3'));
  // });

  // it('if start.month > value.month, set value.month to start.month ', async () => {
  //   const value = new dayjs('1990-01-03');
  //   const start = new dayjs('2019-11-01');
  //   const end = new dayjs('2019-03-01');
  //   const onValueChange = jest.fn();
  //   await createWrapper(start, end, value, onValueChange);
  //   expect(onValueChange).toHaveBeenCalledWith(value.year('2019').month('10'));
  // });

  it('onMonthChange should work correctly', async () => {
    const start = new dayjs('2018-11-01');
    const end = new dayjs('2019-03-01');
    const value = new dayjs('2018-12-03');
    const onValueChange = jest.fn();
    const wrapper = mount(
      {
        render() {
          return (
            <Header
              prefixCls="ant-picker-calendar"
              generateConfig={generateConfig}
              onChange={onValueChange}
              value={value}
              validRange={[start, end]}
              locale={{ year: 'å¹´', locale: 'zh_CN' }}
              mode="month"
            />
          );
        },
      },
      {
        sync: false,
        attachTo: 'body',
      },
    );

    openSelect(wrapper, '.ant-picker-calendar-month-select');
    await sleep(100);
    clickSelectItem(wrapper);
    expect(value.month()).toBe(11);
  });

  it('onTypeChange should work correctly', () => {
    const onTypeChange = jest.fn();
    const value = new dayjs('2018-12-03');
    const wrapper = mount({
      render() {
        return (
          <Header
            prefixCls="ant-picker-calendar"
            generateConfig={generateConfig}
            onModeChange={onTypeChange}
            locale={{ year: 'å¹´', month: 'æ', locale: 'zh_CN' }}
            value={value}
            type="date"
          />
        );
      },
    });
    wrapper.findAll('input[type="radio"]')[1].trigger('change');
    expect(onTypeChange).toHaveBeenCalledWith('year');
  });
});
