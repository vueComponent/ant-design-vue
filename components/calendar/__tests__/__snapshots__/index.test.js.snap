// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar Calendar should support locale 1`] = `
<div class="ant-picker-calendar ant-picker-calendar-full">
  <div class="ant-picker-calendar-header">
    <div class="ant-select ant-picker-calendar-year-select ant-select-single ant-select-show-arrow">
      <!---->
      <!---->
      <div class="ant-select-selector"><span class="ant-select-selection-search"><input type="search" id="rc_select_TEST_OR_SSR" autocomplete="off" class="ant-select-selection-search-input" role="combobox" aria-haspopup="listbox" aria-owns="rc_select_TEST_OR_SSR_list" aria-autocomplete="list" aria-controls="rc_select_TEST_OR_SSR_list" aria-activedescendant="rc_select_TEST_OR_SSR_list_0" readonly="" unselectable="on" style="opacity: 0;"></span><span class="ant-select-selection-item" title="2018年">2018年</span>
        <!---->
      </div><span class="ant-select-arrow" style="user-select: none;" unselectable="on" aria-hidden="true"><span role="img" aria-label="down" class="anticon anticon-down ant-select-suffix"><svg focusable="false" class="" data-icon="down" width="1em" height="1em" fill="currentColor" aria-hidden="true" viewBox="64 64 896 896"><path d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"></path></svg></span>
      <!----></span>
      <!---->
    </div>
    <div class="ant-select ant-picker-calendar-month-select ant-select-single ant-select-show-arrow">
      <!---->
      <!---->
      <div class="ant-select-selector"><span class="ant-select-selection-search"><input type="search" id="rc_select_TEST_OR_SSR" autocomplete="off" class="ant-select-selection-search-input" role="combobox" aria-haspopup="listbox" aria-owns="rc_select_TEST_OR_SSR_list" aria-autocomplete="list" aria-controls="rc_select_TEST_OR_SSR_list" aria-activedescendant="rc_select_TEST_OR_SSR_list_0" readonly="" unselectable="on" style="opacity: 0;"></span><span class="ant-select-selection-item" title="Oct">Oct</span>
        <!---->
      </div><span class="ant-select-arrow" style="user-select: none;" unselectable="on" aria-hidden="true"><span role="img" aria-label="down" class="anticon anticon-down ant-select-suffix"><svg focusable="false" class="" data-icon="down" width="1em" height="1em" fill="currentColor" aria-hidden="true" viewBox="64 64 896 896"><path d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"></path></svg></span>
      <!----></span>
      <!---->
    </div>
    <div class="ant-radio-group ant-radio-group-outline ant-picker-calendar-mode-switch"><label class="ant-radio-button-wrapper ant-radio-button-wrapper-checked"><span class="ant-radio-button ant-radio-button-checked"><input type="radio" class="ant-radio-button-input" value="month"><span class="ant-radio-button-inner"></span></span><span>月</span></label><label class="ant-radio-button-wrapper"><span class="ant-radio-button"><input type="radio" class="ant-radio-button-input" value="year"><span class="ant-radio-button-inner"></span></span><span>年</span></label></div>
  </div>
  <div tabindex="0" class="ant-picker-panel">
    <div class="ant-picker-date-panel">
      <!---->
      <div class="ant-picker-body">
        <table class="ant-picker-content">
          <thead>
            <tr>
              <th>Su</th>
              <th>Mo</th>
              <th>Tu</th>
              <th>We</th>
              <th>Th</th>
              <th>Fr</th>
              <th>Sa</th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td title="2018-09-30" class="ant-picker-cell ant-picker-cell-end">
                <div class="ant-picker-cell-inner ant-picker-calendar-date">
                  <div class="ant-picker-calendar-date-value">30</div>
                  <div class="ant-picker-calendar-date-content">
                    <!---->
                  </div>
                </div>
              </td>
              <td title="2018-10-01" class="ant-picker-cell ant-picker-cell-start ant-picker-cell-in-view">
                <div class="ant-picker-cell-inner ant-picker-calendar-date">
                  <div class="ant-picker-calendar-date-value">01</div>
                  <div class="ant-picker-calendar-date-content">
                    <!---->
                  </div>
                </div>
              </td>
              <td title="2018-10-02" class="ant-picker-cell ant-picker-cell-in-view">
                <div class="ant-picker-cell-inner ant-picker-calendar-date">
                  <div class="ant-picker-calendar-date-value">02</div>
                  <div class="ant-picker-calendar-date-content">
                    <!---->
                  </div>
                </div>
              </td>
              <td title="2018-10-03" class="ant-picker-cell ant-picker-cell-in-view">
                <div class="ant-picker-cell-inner ant-picker-calendar-date">
                  <div class="ant-picker-calendar-date-value">03</div>
                  <div class="ant-picker-calendar-date-content">
                    <!---->
                  </div>
                </div>
              </td>
              <td title="2018-10-04" class="ant-picker-cell ant-picker-cell-in-view">
                <div class="ant-picker-cell-inner ant-picker-calendar-date">
                  <div class="ant-picker-calendar-date-value">04</div>
                  <div class="ant-picker-calendar-date-content">
                    <!---->
                  </div>
                </div>
              </td>
              <td title="2018-10-05" class="ant-picker-cell ant-picker-cell-in-view">
                <div class="ant-picker-cell-inner ant-picker-calendar-date">
                  <div class="ant-picker-calendar-date-value">05</div>
                  <div class="ant-picker-calendar-date-content">
                    <!---->
                  </div>
                </div>
              </td>
              <td title="2018-10-06" class="ant-picker-cell ant-picker-cell-in-view">
                <div class="ant-picker-cell-inner ant-picker-calendar-date">
                  <div class="ant-picker-calendar-date-value">06</div>
                  <div class="ant-picker-calendar-date-content">
                    <!---->
                  </div>
                </div>
              </td>
            </tr>
            <tr>
              <td title="2018-10-07" class="ant-picker-cell ant-picker-cell-in-view">
                <div class="ant-picker-cell-inner ant-picker-calendar-date">
                  <div class="ant-picker-calendar-date-value">07</div>
                  <div class="ant-picker-calendar-date-content">
                    <!---->
                  </div>
                </div>
              </td>
              <td title="2018-10-08" class="ant-picker-cell ant-picker-cell-in-view">
                <div class="ant-picker-cell-inner ant-picker-calendar-date">
                  <div class="ant-picker-calendar-date-value">08</div>
                  <div class="ant-picker-calendar-date-content">
                    <!---->
                  </div>
                </div>
              </td>
              <td title="2018-10-09" class="ant-picker-cell ant-picker-cell-in-view">
                <div class="ant-picker-cell-inner ant-picker-calendar-date">
                  <div class="ant-picker-calendar-date-value">09</div>
                  <div class="ant-picker-calendar-date-content">
                    <!---->
                  </div>
                </div>
              </td>
              <td title="2018-10-10" class="ant-picker-cell ant-picker-cell-in-view">
                <div class="ant-picker-cell-inner ant-picker-calendar-date">
                  <div class="ant-picker-calendar-date-value">10</div>
                  <div class="ant-picker-calendar-date-content">
                    <!---->
                  </div>
                </div>
              </td>
              <td title="2018-10-11" class="ant-picker-cell ant-picker-cell-in-view">
                <div class="ant-picker-cell-inner ant-picker-calendar-date">
                  <div class="ant-picker-calendar-date-value">11</div>
                  <div class="ant-picker-calendar-date-content">
                    <!---->
                  </div>
                </div>
              </td>
              <td title="2018-10-12" class="ant-picker-cell ant-picker-cell-in-view">
                <div class="ant-picker-cell-inner ant-picker-calendar-date">
                  <div class="ant-picker-calendar-date-value">12</div>
                  <div class="ant-picker-calendar-date-content">
                    <!---->
                  </div>
                </div>
              </td>
              <td title="2018-10-13" class="ant-picker-cell ant-picker-cell-in-view">
                <div class="ant-picker-cell-inner ant-picker-calendar-date">
                  <div class="ant-picker-calendar-date-value">13</div>
                  <div class="ant-picker-calendar-date-content">
                    <!---->
                  </div>
                </div>
              </td>
            </tr>
            <tr>
              <td title="2018-10-14" class="ant-picker-cell ant-picker-cell-in-view">
                <div class="ant-picker-cell-inner ant-picker-calendar-date">
                  <div class="ant-picker-calendar-date-value">14</div>
                  <div class="ant-picker-calendar-date-content">
                    <!---->
                  </div>
                </div>
              </td>
              <td title="2018-10-15" class="ant-picker-cell ant-picker-cell-in-view">
                <div class="ant-picker-cell-inner ant-picker-calendar-date">
                  <div class="ant-picker-calendar-date-value">15</div>
                  <div class="ant-picker-calendar-date-content">
                    <!---->
                  </div>
                </div>
              </td>
              <td title="2018-10-16" class="ant-picker-cell ant-picker-cell-in-view">
                <div class="ant-picker-cell-inner ant-picker-calendar-date">
                  <div class="ant-picker-calendar-date-value">16</div>
                  <div class="ant-picker-calendar-date-content">
                    <!---->
                  </div>
                </div>
              </td>
              <td title="2018-10-17" class="ant-picker-cell ant-picker-cell-in-view">
                <div class="ant-picker-cell-inner ant-picker-calendar-date">
                  <div class="ant-picker-calendar-date-value">17</div>
                  <div class="ant-picker-calendar-date-content">
                    <!---->
                  </div>
                </div>
              </td>
              <td title="2018-10-18" class="ant-picker-cell ant-picker-cell-in-view">
                <div class="ant-picker-cell-inner ant-picker-calendar-date">
                  <div class="ant-picker-calendar-date-value">18</div>
                  <div class="ant-picker-calendar-date-content">
                    <!---->
                  </div>
                </div>
              </td>
              <td title="2018-10-19" class="ant-picker-cell ant-picker-cell-in-view ant-picker-cell-today ant-picker-cell-selected">
                <div class="ant-picker-cell-inner ant-picker-calendar-date ant-picker-calendar-date-today">
                  <div class="ant-picker-calendar-date-value">19</div>
                  <div class="ant-picker-calendar-date-content">
                    <!---->
                  </div>
                </div>
              </td>
              <td title="2018-10-20" class="ant-picker-cell ant-picker-cell-in-view">
                <div class="ant-picker-cell-inner ant-picker-calendar-date">
                  <div class="ant-picker-calendar-date-value">20</div>
                  <div class="ant-picker-calendar-date-content">
                    <!---->
                  </div>
                </div>
              </td>
            </tr>
            <tr>
              <td title="2018-10-21" class="ant-picker-cell ant-picker-cell-in-view">
                <div class="ant-picker-cell-inner ant-picker-calendar-date">
                  <div class="ant-picker-calendar-date-value">21</div>
                  <div class="ant-picker-calendar-date-content">
                    <!---->
                  </div>
                </div>
              </td>
              <td title="2018-10-22" class="ant-picker-cell ant-picker-cell-in-view">
                <div class="ant-picker-cell-inner ant-picker-calendar-date">
                  <div class="ant-picker-calendar-date-value">22</div>
                  <div class="ant-picker-calendar-date-content">
                    <!---->
                  </div>
                </div>
              </td>
              <td title="2018-10-23" class="ant-picker-cell ant-picker-cell-in-view">
                <div class="ant-picker-cell-inner ant-picker-calendar-date">
                  <div class="ant-picker-calendar-date-value">23</div>
                  <div class="ant-picker-calendar-date-content">
                    <!---->
                  </div>
                </div>
              </td>
              <td title="2018-10-24" class="ant-picker-cell ant-picker-cell-in-view">
                <div class="ant-picker-cell-inner ant-picker-calendar-date">
                  <div class="ant-picker-calendar-date-value">24</div>
                  <div class="ant-picker-calendar-date-content">
                    <!---->
                  </div>
                </div>
              </td>
              <td title="2018-10-25" class="ant-picker-cell ant-picker-cell-in-view">
                <div class="ant-picker-cell-inner ant-picker-calendar-date">
                  <div class="ant-picker-calendar-date-value">25</div>
                  <div class="ant-picker-calendar-date-content">
                    <!---->
                  </div>
                </div>
              </td>
              <td title="2018-10-26" class="ant-picker-cell ant-picker-cell-in-view">
                <div class="ant-picker-cell-inner ant-picker-calendar-date">
                  <div class="ant-picker-calendar-date-value">26</div>
                  <div class="ant-picker-calendar-date-content">
                    <!---->
                  </div>
                </div>
              </td>
              <td title="2018-10-27" class="ant-picker-cell ant-picker-cell-in-view">
                <div class="ant-picker-cell-inner ant-picker-calendar-date">
                  <div class="ant-picker-calendar-date-value">27</div>
                  <div class="ant-picker-calendar-date-content">
                    <!---->
                  </div>
                </div>
              </td>
            </tr>
            <tr>
              <td title="2018-10-28" class="ant-picker-cell ant-picker-cell-in-view">
                <div class="ant-picker-cell-inner ant-picker-calendar-date">
                  <div class="ant-picker-calendar-date-value">28</div>
                  <div class="ant-picker-calendar-date-content">
                    <!---->
                  </div>
                </div>
              </td>
              <td title="2018-10-29" class="ant-picker-cell ant-picker-cell-in-view">
                <div class="ant-picker-cell-inner ant-picker-calendar-date">
                  <div class="ant-picker-calendar-date-value">29</div>
                  <div class="ant-picker-calendar-date-content">
                    <!---->
                  </div>
                </div>
              </td>
              <td title="2018-10-30" class="ant-picker-cell ant-picker-cell-in-view">
                <div class="ant-picker-cell-inner ant-picker-calendar-date">
                  <div class="ant-picker-calendar-date-value">30</div>
                  <div class="ant-picker-calendar-date-content">
                    <!---->
                  </div>
                </div>
              </td>
              <td title="2018-10-31" class="ant-picker-cell ant-picker-cell-end ant-picker-cell-in-view">
                <div class="ant-picker-cell-inner ant-picker-calendar-date">
                  <div class="ant-picker-calendar-date-value">31</div>
                  <div class="ant-picker-calendar-date-content">
                    <!---->
                  </div>
                </div>
              </td>
              <td title="2018-11-01" class="ant-picker-cell ant-picker-cell-start">
                <div class="ant-picker-cell-inner ant-picker-calendar-date">
                  <div class="ant-picker-calendar-date-value">01</div>
                  <div class="ant-picker-calendar-date-content">
                    <!---->
                  </div>
                </div>
              </td>
              <td title="2018-11-02" class="ant-picker-cell">
                <div class="ant-picker-cell-inner ant-picker-calendar-date">
                  <div class="ant-picker-calendar-date-value">02</div>
                  <div class="ant-picker-calendar-date-content">
                    <!---->
                  </div>
                </div>
              </td>
              <td title="2018-11-03" class="ant-picker-cell">
                <div class="ant-picker-cell-inner ant-picker-calendar-date">
                  <div class="ant-picker-calendar-date-value">03</div>
                  <div class="ant-picker-calendar-date-content">
                    <!---->
                  </div>
                </div>
              </td>
            </tr>
            <tr>
              <td title="2018-11-04" class="ant-picker-cell">
                <div class="ant-picker-cell-inner ant-picker-calendar-date">
                  <div class="ant-picker-calendar-date-value">04</div>
                  <div class="ant-picker-calendar-date-content">
                    <!---->
                  </div>
                </div>
              </td>
              <td title="2018-11-05" class="ant-picker-cell">
                <div class="ant-picker-cell-inner ant-picker-calendar-date">
                  <div class="ant-picker-calendar-date-value">05</div>
                  <div class="ant-picker-calendar-date-content">
                    <!---->
                  </div>
                </div>
              </td>
              <td title="2018-11-06" class="ant-picker-cell">
                <div class="ant-picker-cell-inner ant-picker-calendar-date">
                  <div class="ant-picker-calendar-date-value">06</div>
                  <div class="ant-picker-calendar-date-content">
                    <!---->
                  </div>
                </div>
              </td>
              <td title="2018-11-07" class="ant-picker-cell">
                <div class="ant-picker-cell-inner ant-picker-calendar-date">
                  <div class="ant-picker-calendar-date-value">07</div>
                  <div class="ant-picker-calendar-date-content">
                    <!---->
                  </div>
                </div>
              </td>
              <td title="2018-11-08" class="ant-picker-cell">
                <div class="ant-picker-cell-inner ant-picker-calendar-date">
                  <div class="ant-picker-calendar-date-value">08</div>
                  <div class="ant-picker-calendar-date-content">
                    <!---->
                  </div>
                </div>
              </td>
              <td title="2018-11-09" class="ant-picker-cell">
                <div class="ant-picker-cell-inner ant-picker-calendar-date">
                  <div class="ant-picker-calendar-date-value">09</div>
                  <div class="ant-picker-calendar-date-content">
                    <!---->
                  </div>
                </div>
              </td>
              <td title="2018-11-10" class="ant-picker-cell">
                <div class="ant-picker-cell-inner ant-picker-calendar-date">
                  <div class="ant-picker-calendar-date-value">10</div>
                  <div class="ant-picker-calendar-date-content">
                    <!---->
                  </div>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
    <!---->
  </div>
</div>
`;
